### ISSUE TITLE
PHP Language Server indexes files strangely

### ISSUE THREAD

**Issue by d7sd6u** [kak-lsp/kak-lsp#420]
Hey. I am new to Kakoune and wanted to use it with PHP code to learn a little about it. I tried to set up kak-lsp with PHP Language Server but came across strange behavior. When lsp indexes files, it indexes all files in the project except the file in which indexing started. For example, if I open the file x.php, then LSP indexes the files y.php, z.php, w.php and others, but not the x.php file. Accordingly, the commands lsp-definition, lsp-references and others work fine for functions and classes from files y.php, z.php, w.php and others, but not from x.php.

You can reproduce this behavior for example in the [rss-bridge](https://github.com/RSS-Bridge/rss-bridge) repo, try opening one of the files in the bridges/ folder there and trigger indexing.

I installed kak-lsp manually as described in the README:
```shell
wget https://github.com/kak-lsp/kak-lsp/releases/download/v8.0.0/kak-lsp-v8.0.0-x86_64-unknown-linux-musl.tar.gz`
tar xzvf kak-lsp-v8.0.0-x86_64-unknown-linux-musl.tar.gz

# replace `~/.local/bin/` with something on your `$PATH`
mv kak-lsp ~/.local/bin/

mkdir -p ~/.config/kak-lsp
mv kak-lsp.toml ~/.config/kak-lsp/`
```
And installed PHP Language Server like this:
```shell
composer global require jetbrains/phpstorm-stubs:dev-master
composer global require require felixfbecker/language-server
cd ~/.config/composer
rm vendor/jetbrains/phpstorm-stubs/http/http.php # workaround for this issue: https://github.com/felixfbecker/php-language-server/issues/795
composer run-script --working-dir=vendor/felixfbecker/language-server parse-stubs
```
I am using Linux Mint 20 and Kakoune was installed from Debian repositories.
I have attached my kakrc, kak-lsp.toml and debug buffer:
[kakrc](https://github.com/kak-lsp/kak-lsp/files/5617453/kakrc.txt)
[debug-buffer.log](https://github.com/kak-lsp/kak-lsp/files/5617449/debug-buffer.log)
[kak-lsp.toml](https://github.com/kak-lsp/kak-lsp/files/5617462/kak-lsp.toml.txt)

What am I doing wrong?
P.S. Sorry for my bad english

**Comment by krobelus** [comment]
Thanks for the nice report! I can reproduce your observation that there is no line like `LSP: Parsing file ...` for the current buffer's file. However, everything seems to be working fine regardless.

For example if I
1. start Kakoune in the rss-bridge repo as `kak lib/error.php`
2. move to `function returnError` 
3. run `:lsp-references`

I get this result, containing references from inside `lib/error.php` and from elsewhere.

```
actions/ConnectivityAction.php:28:4:			returnError('This action is only available in debug mode!');
lib/contents.php:200:4:			returnError(<<<EOD
lib/error.php:33:2:	returnError($message, 400);
lib/error.php:42:2:	returnError($message, 500);
```

I believe that whenever the current buffer is modified, kak-lsp sends the buffer contents to the language server. Hence there is no need to parse the on-disk file. This allows to use all LSP features before actually saving the buffer to a file.

If that doesn't help, can you give specific steps that don't work as expected?

**Comment by d7sd6u** [comment]
> Thanks for the nice report! I can reproduce your observation that there is no line like `LSP: Parsing file ...` for the current buffer's file. However, everything seems to be working fine regardless.
> 
> For example if I
> 
>     1. start Kakoune in the rss-bridge repo as `kak lib/error.php`
> 
>     2. move to `function returnError`
> 
>     3. run `:lsp-references`
> 
> 
> I get this result, containing references from inside `lib/error.php` and from elsewhere.
> 
Yes. I have the same behavior in `lib/error.php`. Apparently the bug appears only in php classes. Try to open the file with php class:
```shell
kak bridge/VkBridge.php
```
Move to `$this->getPhoto` and run `:lsp-references` or `:lsp-definition`, nothing happens.
Without closing Kakoune, open another similar file: `:e bridges/FacebookBridge.php`
Move to `$this->collectGroupData` and run `:lsp-references` or `:lsp-definition`, and everything works!
You can go back to the previous file and it still won't work.
Alternatively you can close Kakoune and open `FacebookBridge.php` first and everything will be the other way around, `:lsp-definition` won't work in `FacebookBridge.php` but will work in `VkBridge.php` and all other files.

**Comment by krobelus** [comment]
OK, I can reproduce. Curiously, everything works when running `lsp-enable` **after** opening the file, so remove lsp-enable from your config and run

```
kak -e 'edit bridges/VkBridge.php 171 22; lsp-enable'
```

It feels like this could really be an issue in kak-lsp.

I use a keybinding to run `lsp-enable` because I mostly use LSP features in long-running sessions, no need to always run it.
Some people use `WinSetOption filetype=php` style hooks to run `lsp-enable-window` conditionally.

**Comment by d7sd6u** [comment]
> Curiously, everything works when running lsp-enable after opening the file, so remove lsp-enable from your config and run

@krobelus, but it doesn't work for me (I removed lsp-enable from kakrc). And if you manually open the file and then run `lsp-enable` or `lsp-enable-window` or with `WinSetOption filetype = php`, it also does not work (technically it works but with the above problem). But if you remove `lsp-enable`, manually open the file, DO NOT run `lsp-enable` and run `lsp-definition` (or `lsp-reference` for example) on some function, then everything works! Yes, for functions from external files for the first time, you need to wait 2-3 seconds for indexing, but at least for functions defined inside the file, you hardly need to wait (less than one second). This is not a very elegant solution, but it suits me.
